alf/algorithms/td_loss.py

```diff
@@ -106,15 +113,30 @@ def forward(self, experience, value, target_value):
         if self._debug_summaries and alf.summary.should_record_summaries():
             mask = experience.step_type[:-1] != StepType.LAST
             with alf.summary.scope(self._name):
-                alf.summary.scalar(
-                    "explained_variance_of_return_by_value",
-                    tensor_utils.explained_variance(value, returns, mask))
-                safe_mean_hist_summary('values', value, mask)
-                safe_mean_hist_summary('returns', returns, mask)
-                safe_mean_hist_summary("td_error", returns - value, mask)
+
+                def _summarize(v, r, td, suffix):
+                    alf.summary.scalar(
+                        "explained_variance_of_return_by_value" + suffix,
+                        tensor_utils.explained_variance(v, r, mask))
+                    safe_mean_hist_summary('values' + suffix, v, mask)
+                    safe_mean_hist_summary('returns' + suffix, r, mask)
+                    safe_mean_hist_summary("td_error" + suffix, td, mask)
+
+                if value.ndim == 2:
+                    _summarize(value, returns, returns - value, '')
+                else:
+                    td = returns - value
+                    for i in range(value.shape[2]):
+                        suffix = '/' + str(i)
+                        _summarize(value[..., i], returns[..., i], td[..., i],
+                                   suffix)
 
         loss = self._td_error_loss_fn(returns.detach(), value)
 
+        if loss.ndim == 3:
+            # Multidimensional reward. Average over the critic loss for all dimensions
+            loss = loss.mean(dim=2)
```

Comment:
The current implementation of critic network trains all rewards together. Even if we have separate critic network for each reward, mean() or sum() doesn't matter much because Adam optimizer is largely invariant of it.